boot/freeldr/freeldr/lib/hwide.c

```diff
@@ -0,0 +1,761 @@
+static BOOLEAN SelectDevice(UCHAR Channel, UCHAR DeviceNumber);
+static BOOLEAN IdentifyDevice(UCHAR Channel, UCHAR DeviceNumber, PDEVICE_UNIT *DeviceUnit);
+static BOOLEAN AtapiReadyCheck(PDEVICE_UNIT DeviceUnit);
+static BOOLEAN AtapiReadLogicalSectorLBA(PDEVICE_UNIT DeviceUnit, ULONGLONG SectorNumber, PVOID Buffer);
+static BOOLEAN AtaReadLogicalSectorsLBA(PDEVICE_UNIT DeviceUnit, ULONGLONG SectorNumber, ULONG SectorCount, PVOID Buffer);
+
+/* FUNCTIONS ******************************************************************/
+
+/* Don't call this before running the system timer calibration */
+BOOLEAN AtaInit(PUCHAR DetectedCount)
+{
+    UCHAR Channel, DeviceNumber;
+    PDEVICE_UNIT DeviceUnit = NULL;
+
+    TRACE("AtaInit()\n");
+
+    *DetectedCount = 0;
+
+    RtlZeroMemory(&Units, sizeof(Units));
+
+    /* Detect and enumerate ATA/ATAPI devices */
+    for (Channel = 0; Channel < MAX_CHANNELS; Channel++)
+    {
+        for (DeviceNumber = 0; DeviceNumber < MAX_DEVICES; DeviceNumber++)
+        {
+            if (IdentifyDevice(Channel, DeviceNumber, &DeviceUnit))
+            {
+                Units[(*DetectedCount)++] = DeviceUnit;
+            }
+        }
+    }
+
+    if (*DetectedCount > 0)
+        return TRUE;
+    else
+        return FALSE;
+}
+
+VOID
+AtaFree(VOID)
+{
+    UCHAR i;
+
+    TRACE("AtaFree()\n");
+
+    for (i = 0; i < sizeof(Units) / sizeof(PDEVICE_UNIT); i++)
+    {
+        if(Units[i])
+            FrLdrTempFree(Units[i], TAG_ATA_DEVICE);
+    }
+}
+
+PDEVICE_UNIT
+AtaGetDevice(UCHAR UnitNumber)
+{
+    if (UnitNumber < MAX_CHANNELS * MAX_DEVICES)
+        return Units[UnitNumber];
+    else
+        return NULL;
+}
+
+BOOLEAN
+AtaAtapiReadLogicalSectorsLBA(PDEVICE_UNIT DeviceUnit, ULONGLONG SectorNumber, ULONG SectorCount, PVOID Buffer)
+{
+    UCHAR RetryCount;
+    BOOLEAN Status;
+
+    if (!DeviceUnit)
+        return FALSE;
+
+    if (DeviceUnit->Flags & ATA_DEVICE_ATAPI)
+    {
+        if ((DeviceUnit->Flags & ATA_DEVICE_NO_MEDIA) || (DeviceUnit->Flags & ATA_DEVICE_NOT_READY))
+        {
+            /* Retry 4 times */
+            for (RetryCount = 0; RetryCount < 4; RetryCount++)
+            {
+                /* Make the device ready */
+                if (AtapiReadyCheck(DeviceUnit))
+                    break;
+            }
+            if (RetryCount >= 4)
+            {
+                ERR("AtaAtapiReadLogicalSectorsLBA() Device not ready.\n");
+                return FALSE;
+            }
+        }
+        if (SectorNumber + SectorCount > DeviceUnit->TotalSectors + 1)
+        {
+            ERR("AtaAtapiReadLogicalSectorsLBA() Attempt to read more than there is to read.\n");
+            return FALSE;
+        }
+
+        do
+        {
+            /* Read a single sector */
+            Status = AtapiReadLogicalSectorLBA(DeviceUnit, SectorNumber, Buffer);
+            if (!Status)
+                return FALSE;
+
+            SectorCount--;
```

Comment:
>A value of 00h indicates that 256 sectors are to be transferred

Reply:
Is it wanted or not? Because by looking at the code I see that the do-loop starts with SectorCount == 0, and then this decrementing (since SectorCount is a ULONG) will make it equal to 0xFFFFFFFF == 4294967295, and the condition in the while will succeed, etc...

Reply:
You're right. Added check to avoid this case.